File: watcher-frontend/src/types/presupuesto.ts
export interface Programa {
  id: number;
  ejercicio: number;
  organismo: string;
  programa: string;
  subprograma?: string;
  partida_presupuestaria: string;
  descripcion: string;
  monto_inicial: number;
  monto_vigente: number;
  fecha_aprobacion: string;
  meta_fisica?: string;
  meta_numerica?: number;
  unidad_medida?: string;
  fuente_financiamiento?: string;
}

export interface Ejecucion {
  id: number;
  boletin_id?: number;
  presupuesto_base_id?: number;
  fecha_boletin: string;
  organismo?: string;
  beneficiario?: string;
  concepto?: string;
  monto: number | string;
  tipo_operacion?: string;
  partida_presupuestaria?: string;
  programa?: string;
  categoria_watcher?: string;
  riesgo_watcher?: string;
  etapa_gasto?: string;
  jurisdiccion?: string;
  monto_acumulado_mes?: number | string;
  monto_acumulado_trimestre?: number | string;
  monto_acumulado_anual?: number | string;
  requiere_revision?: boolean;
  is_duplicate: number;
  observaciones?: string;
}

export interface EjecucionListResponse {
  ejecuciones: Ejecucion[];
  total: number;
  total_monto: number | string;
  page: number;
  page_size: number;
}

export interface OrgResumenItem {
  organismo: string | null;
  count: number;
  monto_total: number | string;
  monto_compromiso: number | string;
  monto_ejecucion: number | string;
  monto_vigente: number | string | null;
  pct_compromiso: number | null;
  pct_ejecucion: number | null;
  sobre_compromiso: boolean;
  sobre_ejecucion: boolean;
  matched: boolean;
  /** Portion of monto_compromiso that is only a tender call (intención). */
  monto_llamado: number | string;
}

export interface MesResumenItem {
  mes: string; // "2026-02", "2026-03"
  count: number;
  monto_total: number | string;
}

export interface CoberturaResumen {
  monto_total: number | string;
  monto_con_denominador: number | string;
  monto_sin_denominador: number | string;
  count_sin_denominador: number;
  pct_sin_denominador: number;
}

export interface CoberturaTemporalResumen {
  mes_desde: string | null;
  mes_hasta: string | null;
  meses_cubiertos: number;
  meses_del_ejercicio: number;
  /** Months already elapsed with no ingestion — a real gap. */
  meses_vencidos_sin_ingesta: string[];
  /** Months of the ejercicio that have not happened yet. */
  meses_futuros: string[];
  dias_con_publicacion: number;
  dias_justificados: number;
  dias_faltantes: number;
  denominador_es_anual: boolean;
}

export interface DenominadorSinDuenoItemResumen {
  organismo: string;
  count: number;
  monto_vigente: number | string;
}

export interface DenominadorSinDuenoResumen {
  monto_total: number | string;
  monto_sin_dueno: number | string;
  monto_verificable: number | string;
  count_sin_dueno: number;
  pct_sin_dueno: number;
  por_organismo: DenominadorSinDuenoItemResumen[];
}

export interface EjecucionResumenResponse {
  total_canonical: number;
  total_duplicates: number;
  monto_canonical: number | string;
  monto_duplicates: number | string;
  monto_compromiso: number | string;
  monto_ejecucion: number | string;
  sobre_compromiso_count: number;
  cobertura: CoberturaResumen;
  cobertura_temporal: CoberturaTemporalResumen;
  denominador: DenominadorSinDuenoResumen;
  por_organismo: OrgResumenItem[];
  por_mes: MesResumenItem[];
}

export type JurisdiccionGasto = "provincial" | "municipal" | "fuera_presupuesto"
export type SerieGasto = "compromiso" | "ejecucion"

export interface EjecucionFilters {
  skip?: number;
  limit?: number;
  fecha_desde?: string;
  fecha_hasta?: string;
  organismo?: string;
  riesgo?: string;
  solo_canonicos?: boolean;
  presupuesto_base_id?: number;
  requiere_revision?: boolean;
  jurisdiccion?: JurisdiccionGasto;
  etapa_gasto?: string;
}

export interface ProgramaDetail extends Programa {
  ejecuciones: Ejecucion[];
  total_ejecutado: number;
  porcentaje_ejecucion: number;
}

export interface PresupuestoFilters {
  skip?: number;
  limit?: number;
  ejercicio?: number;
  organismo?: string;
}

export interface ProgramasListResponse {
  programas: Programa[];
  total: number;
  page: number;
  page_size: number;
}

export interface Organismo {
  organismo: string;
  total_programas: number;
  monto_inicial_total: number;
  monto_vigente_total: number;
}

File: watcher-backend/app/schemas/presupuesto.py
"""
Schemas for Presupuesto endpoints
"""

from datetime import date
from typing import Annotated

from pydantic import BaseModel, PlainSerializer


# ARS values are emitted in millions so JSON never contains 11+ digit
# integers (those get corrupted by some browser inspectors and axios then
# swallows the parse error, which made the UI look empty).
def _money_millions(value: float) -> float:
    return round(float(value) / 1_000_000.0, 4)


JsonMoney = Annotated[
    float,
    PlainSerializer(_money_millions, return_type=float, when_used="json"),
]


class ProgramaBase(BaseModel):
    ejercicio: int
    organismo: str
    programa: str
    subprograma: str | None = None
    partida_presupuestaria: str
    descripcion: str
    monto_inicial: JsonMoney
    monto_vigente: JsonMoney

class ProgramaResponse(ProgramaBase):
    id: int
    fecha_aprobacion: date
    meta_fisica: str | None = None
    meta_numerica: float | None = None
    unidad_medida: str | None = None
    fuente_financiamiento: str | None = None

    class Config:
        from_attributes = True

class EjecucionResponse(BaseModel):
    id: int
    boletin_id: int | None = None
    presupuesto_base_id: int | None = None
    fecha_boletin: date
    organismo: str | None = None
    beneficiario: str | None = None
    concepto: str | None = None
    monto: JsonMoney
    tipo_operacion: str | None = None
    partida_presupuestaria: str | None = None
    programa: str | None = None
    categoria_watcher: str | None = None
    riesgo_watcher: str | None = None
    etapa_gasto: str | None = None
    jurisdiccion: str | None = None
    monto_acumulado_mes: JsonMoney | None = None
    monto_acumulado_trimestre: JsonMoney | None = None
    monto_acumulado_anual: JsonMoney | None = None
    requiere_revision: bool | None = False
    is_duplicate: int = 0
    observaciones: str | None = None

    class Config:
        from_attributes = True


class EjecucionListResponse(BaseModel):
    ejecuciones: list[EjecucionResponse]
    total: int
    total_monto: JsonMoney
    page: int
    page_size: int


class OrgResumenItem(BaseModel):
    organismo: str | None
    count: int
    monto_total: JsonMoney
    monto_compromiso: JsonMoney = 0.0
    monto_ejecucion: JsonMoney = 0.0
    monto_vigente: JsonMoney | None = None
    pct_compromiso: float | None = None
    pct_ejecucion: float | None = None
    sobre_compromiso: bool = False
    sobre_ejecucion: bool = False
    matched: bool = False


class MesResumenItem(BaseModel):
    mes: str   # "2026-02", "2026-03", ...
    count: int
    monto_total: JsonMoney


class EjecucionResumenResponse(BaseModel):
    total_canonical: int
    total_duplicates: int
    monto_canonical: JsonMoney
    monto_duplicates: JsonMoney
    monto_compromiso: JsonMoney = 0.0
    monto_ejecucion: JsonMoney = 0.0
    sobre_compromiso_count: int = 0
    por_organismo: list[OrgResumenItem]
    por_mes: list[MesResumenItem]

class ProgramaDetailResponse(ProgramaResponse):
    ejecuciones: list[EjecucionResponse]
    total_ejecutado: JsonMoney
    porcentaje_ejecucion: float

class ProgramasListResponse(BaseModel):
    programas: list[ProgramaResponse]
    total: int
    page: int
    page_size: int

class OrganismoResponse(BaseModel):
    organismo: str
    total_programas: int
    monto_inicial_total: JsonMoney
    monto_vigente_total: JsonMoney

